Add support for JPA attribute converters and field annotations

aldenquimby opened the pull request:
#### Summary
1. Adds `AttributeConverter`s to object loaders, which lets users specify conversions of certain types before appending each field to an infile row
  - my use case was mostly joda time and enums
  - generic solution allows anyone to plug in additional types
2. Update annotation inspector to allow JPA annotations to be on fields instead of getters

aperezlo commented on src/main/java/com/opower/persistence/jpile/loader/SingleInfileObjectLoader.java:
```diff
@@ -175,4 +186,18 @@ boolean isAutoGenerateId() {
+     * 
+     * @author aldenquimby@gmail.com
+     */
+    public interface FieldConverter<T> {
```
What's the benefit of defining your own interface here as compared to just using Google's ``Function`` interface?

aldenquimby replied:
Main benefit would be readability, so it's clear that some implementation
is meant to convert a particular field before loading it into the db. If
you don't think it's necessary, a Function would work just as well.

On Monday, February 9, 2015, Andrew Perez-Lopez <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> In
> src/main/java/com/opower/persistence/jpile/loader/SingleInfileObjectLoader.java
> <https://github.com/opower/jpile/pull/35#discussion_r24347096>:
>
> > @@ -175,4 +186,18 @@ boolean isAutoGenerateId() {
> >      InfileDataBuffer getInfileDataBuffer() {
> >          return infileDataBuffer;
> >      }
> > +
> > +    /**
> > +     * An interface for converting particular fields into objects used by {@link InfileRow}.
> > +     *
> > +     * @author aldenquimby@gmail.com
> > +     */
> > +    public interface FieldConverter<T> {
>
> What's the benefit of defining your own interface here as compared to just
> using Google's Function interface?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opower/jpile/pull/35/files#r24347096>.
>

s-machina replied:
I agree with @aperezlo. I think your use is consistent with the goals of <code>Function</code>, even though it's unlikely that you'll be transforming full collections of one type.

aldenquimby replied:
It gets a bit ugly to have `Map<Class<?>, Function<?, ?>>` because I would need to cast to `Function<Object, Object>` and suppress an unchecked warning after getting from the map. The FieldConverter interface doesn't have the same problem.

aperezlo replied:
How about at least making your interface extend the ``Function`` interface?  I.e.
```java
public interface FieldConverter<T> extends Function<T, Object> {
...
```
then you get the convenience of referring to a ``FieldConverter``, but are also not entirely re-inventing the wheel.

karkum commented on src/main/java/com/opower/persistence/jpile/reflection/PersistenceAnnotationInspector.java:
```diff
@@ -57,7 +57,14 @@
      * @return the annotation on the method or null if it doesn't exist
      */
     public <A extends Annotation> A findAnnotation(Method method, Class<A> annotationType) {
```
Can you please update the javadoc for this method?

karkum commented on src/main/java/com/opower/persistence/jpile/loader/SingleInfileObjectLoader.java:
```diff
@@ -75,7 +77,11 @@ public void convertToInfileRow(E entity, InfileRow infileRow) {
+                    if (converter != null) {
+                        infileRow.append(converter.convert(object));
+                    }
+                    else if (object instanceof Date) {
```
Should this be be just an ```if```? What if we have a ```FieldConverter``` and the objet is an instance of a Date. They seem like distinct conditions.

karkum commented:
@aldenquimby there are a few check style problems with this, could you please fix them?

    jpile/src/main/java/com/opower/persistence/jpile/loader/SingleInfileObjectLoaderBuilder.java:71: Line is longer than 128 characters (found 146).

    jpile/src/main/java/com/opower/persistence/jpile/reflection/PersistenceAnnotationInspector.java:272:28: '||' should be on a new line.

    jpile/src/main/java/com/opower/persistence/jpile/reflection/PersistenceAnnotationInspector.java:273:44: '||' should be on a new line

    jpile/src/main/java/com/opower/persistence/jpile/reflection/PersistenceAnnotationInspector.java:274:51: '||' should be on a new line.

    jpile/src/test/java/com/opower/persistence/jpile/infile/driver/HikariJdbcDriverSupportTest.java:19: Javadoc has empty description section.

    jpile/src/test/java/com/opower/persistence/jpile/infile/driver/HikariJdbcDriverSupportTest.java:33:17: Name 'testAccept_hikariStatement_returnsTrue' must match pattern '^[a-z][a-zA-Z0-9]*$'.

    jpile/src/test/java/com/opower/persistence/jpile/infile/driver/HikariJdbcDriverSupportTest.java:38:17: Name 'testAccept_mysqlStatement_returnsFalse' must match pattern '^[a-z][a-zA-Z0-9]*$'

    jpile/src/test/java/com/opower/persistence/jpile/infile/driver/HikariJdbcDriverSupportTest.java:43:17: Name 'testDoWithStatement_unwrapsStatement_setsLocalInfileInputStream' must match pattern '^[a-z][a-zA-Z0-9]*$'.

    jpile/src/test/java/com/opower/persistence/jpile/sample/Phone.java:3: Using the '.*' form of import should be avoided - javax.persistence.*.

karkum commented:
There are two warnings as well:

    [WARNING] jpile/src/main/java/com/opower/persistence/jpile/loader/SingleInfileObjectLoader.java:[82,59] unchecked call to convert(T) as a member of the raw type com.opower.persistence.jpile.loader.SingleInfileObjectLoader.FieldConverter
    [WARNING] jpile/src/main/java/com/opower/persistence/jpile/loader/SingleInfileObjectLoader.java:[124,37] unchecked call to convert(T) as a member of the raw type com.opower.persistence.jpile.loader.SingleInfileObjectLoader.FieldConverter

aldenquimby commented:
@karkum Thanks for the review, I'll update things based on your comments. Do you know if CI is setup somewhere for this project? Would be great to use something like https://travis-ci.org and have the build fail if check style fails.

karkum commented on src/main/java/com/opower/persistence/jpile/loader/SingleInfileObjectLoader.java:
```diff
@@ -113,6 +119,11 @@ else if (object.getClass().isEnum()) {
      */
     @VisibleForTesting
     Object getEnumValueToAppend(Method method, Enum<?> enumObject) {
+        FieldConverter converter = fieldConverters.get(Enum.class);
```
Is this necessary? Would it not be handled by line 80?

I added a test locally (added an enum to one of our model objects and added a FieldConverter for that enum) and it did not reach this point and instead converted the value and appended it to the infile ro on line 82.

aldenquimby replied:
I added this to handle the use case of sharing a FieldConverter for all enums, sort of like the JPA `@Enumerated` thing with `EnumType`. Rather than registering a field converter for each enum class in an application, registering it for `Enum.class` means it will be applied to all enums. Does that make sense? I wasn't sure how else to handle that.

karkum replied:
Ah I see. Is that a use case you have currently? If so, we should add a test for it. If not, it may not be [something that is really necessary](http://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

karkum commented:
@aldenquimby, we do have CI for this project but it is internal to our organization. I'll ask around and see if we can set something up for our open source projects.

karkum commented:
@aldenquimby any updates on this?

dean-allen commented:
I'm a bit late to this PR, but thought I'd ask anyway:  Is there any plan to upgrade the JPA dependency to 2.1?  If so, that spec already includes the concept of converters.  See https://docs.jboss.org/hibernate/jpa/2.1/api/ (specifically the `AttributeConverter`, `Converts`, `Converter`, and `Convert` classes).  Should we upgrade and provide support for that functionality?

aldenquimby commented:
Hi all, sorry for the delayed response here.

JPA 2.1s `AttributeConverter` looks like exactly what I was looking for with the field converters. If everyone is ok with upgrading to 2.1, I think that makes the most sense.

What do you guys think? Next best option seems to be `public interface FieldConverter<T> extends Function<T, Object>`

karkum commented:
@aldenquimby, I think that makes sense.

aldenquimby commented:
@karkum I upgraded to JPA 2.1, replaced `FieldConverter` with `AttributeConverter`, fixed the checkStyle issues you listed, and added a test for a registered enum converter. Is there anything else I should do for this PR?

karkum commented:
@aldenquimby, Sorry for the delayed response. There's still one more checkstyle issue:

    jpile/src/test/java/com/opower/persistence/jpile/loader/SingleInfileObjectLoaderBuilderTest.java:195:22: Name 'PREFIX' must match pattern '^[a-z][a-zA-Z0-9]*$'.

I will work on getting CI for this project, so we don't have to go through this again

daviddannunzio commented:
@karkum definitely forgot I was still getting notifications for this repo......Glad to see you're still watching over it! :smiley:

karkum commented:
@daviddannunzio :+1: your input is welcome as well!

karkum commented on src/main/java/com/opower/persistence/jpile/loader/HierarchicalInfileObjectLoader.java:
```diff
@@ -339,6 +343,10 @@ public void setUseReplace(boolean useReplace) {
         this.useReplace = useReplace;
     }
 
+    public void registerAttributeConverters(Map<Class<?>, AttributeConverter<?, ?>> converters) {
```
Can you please add javadoc here to specify how one would use this?

karkum commented:
@aldenquimby, this looks good to merge. Let me know when you have fixed the checkstyle and added that javadoc

karkum commented:
@aldenquimby, we have also added CI for this project, see https://travis-ci.org/opower/jpile

If you add another commit, we should be able to see any failures prior to merging

karkum commented:
ping @aldenquimby

amir20 commented:
@karkum let's close this until @aldenquimby responds.
